# Tunning RandomForest

## 1 Xử lí dữ liệu

### 1.1 Thư viện

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import IPython.display as display
import math

### 1.2 Đọc tập dữ liệu 

In [2]:
data_path  = "../../data/pima-indians-diabetes.csv"
data_names = ["Pregnant","2Hr_Gluco_Concentrate","Blood_Pressure","Skin_Fold","2Hr_Isnulin","BMI","Predigy","Age","Class"]
df = pd.read_csv(data_path, names = data_names)

In [3]:
df.describe()

Pregnant  2Hr_Gluco_Concentrate  Blood_Pressure   Skin_Fold  \
count  768.000000             768.000000      768.000000  768.000000   
mean     3.845052             120.894531       69.105469   20.536458   
std      3.369578              31.972618       19.355807   15.952218   
min      0.000000               0.000000        0.000000    0.000000   
25%      1.000000              99.000000       62.000000    0.000000   
50%      3.000000             117.000000       72.000000   23.000000   
75%      6.000000             140.250000       80.000000   32.000000   
max     17.000000             199.000000      122.000000   99.000000   

       2Hr_Isnulin         BMI     Predigy         Age       Class  
count   768.000000  768.000000  768.000000  768.000000  768.000000  
mean     79.799479   31.992578    0.471876   33.240885    0.348958  
std     115.244002    7.884160    0.331329   11.760232    0.476951  
min       0.000000    0.000000    0.078000   21.000000    0.000000  
25%       0.000000   27.300000    0.243750   24.000000    0.000000  
50%      30.500000   32.000000    0.372500   29.000000    0.000000  
75%     127.250000   36.600000    0.626250   41.000000    1.000000  
max     846.000000   67.100000    2.420000   81.000000    1.000000

### 1.3 Xử lí dữ liệu thiếu

In [4]:
df["2Hr_Gluco_Concentrate"] = df["2Hr_Gluco_Concentrate"].replace(0, np.nan)
df["2Hr_Gluco_Concentrate"] = df["2Hr_Gluco_Concentrate"].fillna(df["2Hr_Gluco_Concentrate"].median())
df["Blood_Pressure"] = df["Blood_Pressure"].replace(0, np.nan)
df["Blood_Pressure"] = df.groupby("Class")["Blood_Pressure"].transform(lambda x: x.fillna(x.median()))
df["Skin_Fold"] = df["Skin_Fold"].replace(0, np.nan)
df["Skin_Fold"] = df.groupby("Class")["Skin_Fold"].transform(lambda x: x.fillna(x.median()))
df["Skin_Fold"] = df["Skin_Fold"].replace(0, np.nan)
df["Skin_Fold"] = df.groupby("Class")["Skin_Fold"].transform(lambda x: x.fillna(x.median()))
df["2Hr_Isnulin"] = df["2Hr_Isnulin"].replace(0, np.nan)
df["2Hr_Isnulin"] = df.groupby("Class")["2Hr_Isnulin"].transform(lambda x: x.fillna(x.median()))
df["BMI"] = df["BMI"].replace(0, np.nan)
df["BMI"] = df.groupby("Class")["BMI"].transform(lambda x: x.fillna(x.median()))
df.describe()

Pregnant  2Hr_Gluco_Concentrate  Blood_Pressure   Skin_Fold  \
count  768.000000             768.000000      768.000000  768.000000   
mean     3.845052             121.656250       72.389323   29.089844   
std      3.369578              30.438286       12.106039    8.890820   
min      0.000000              44.000000       24.000000    7.000000   
25%      1.000000              99.750000       64.000000   25.000000   
50%      3.000000             117.000000       72.000000   28.000000   
75%      6.000000             140.250000       80.000000   32.000000   
max     17.000000             199.000000      122.000000   99.000000   

       2Hr_Isnulin         BMI     Predigy         Age       Class  
count   768.000000  768.000000  768.000000  768.000000  768.000000  
mean    141.753906   32.434635    0.471876   33.240885    0.348958  
std      89.100847    6.880498    0.331329   11.760232    0.476951  
min      14.000000   18.200000    0.078000   21.000000    0.000000  
25%     102.500000   27.500000    0.243750   24.000000    0.000000  
50%     102.500000   32.050000    0.372500   29.000000    0.000000  
75%     169.500000   36.600000    0.626250   41.000000    1.000000  
max     846.000000   67.100000    2.420000   81.000000    1.000000

- Thực hiện xử lí dữ liệu thiếu theo trung vị của lớp output

### 1.4 Spilt Train-Test

In [5]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from numpy import set_printoptions
# separate array into input and output components
Y = df["Class"]
X = df.drop("Class",axis=1)
seed = 7
validation_size = 0.2
X_train, X_validation, Y_train, Y_validation = train_test_split(X, Y, test_size=validation_size, random_state=seed)

### 1.5 Normalization & Standardlize

In [6]:
scaler = MinMaxScaler(feature_range=(0, 1))
X_train = scaler.fit_transform(X_train)
X_validation = scaler.transform(X_validation)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_validation = scaler.transform(X_validation)
set_printoptions(precision=3)
print(X_train)

[[ 3.548e-01 -4.247e-01 -1.018e-03 ...  5.072e-01 -6.179e-01  1.064e-02]
 [-5.479e-01 -6.220e-01  1.166e+00 ...  1.856e+00 -1.035e+00 -8.605e-01]
 [ 5.391e-02  1.219e+00  4.993e-01 ...  4.800e-02  1.037e+00 -1.636e-01]
 ...
 [-1.150e+00 -2.102e+00 -1.002e+00 ... -1.559e+00  8.283e-01  2.972e+00]
 [-8.489e-01 -5.233e-01 -1.168e+00 ... -1.186e+00 -8.507e-01 -1.035e+00]
 [ 1.258e+00  1.910e+00 -1.018e-03 ...  1.930e-02  7.793e-01  2.720e-01]]


- Rescale các giá trị số về [0,1] và theo phân bố chuẩn

### 1.6 PCA
- Thử giữ số phương sai cho PCA là 95%

In [7]:
from sklearn.decomposition import PCA
pca = PCA(n_components=0.95)
X_train = pca.fit_transform(X_train)
X_validation = pca.transform(X_validation)

## 2. Tunning (Grid_Search)

### 2.1 Tunning (GridSearch)

In [8]:
from sklearn.model_selection import GridSearchCV, KFold
from sklearn.ensemble import RandomForestClassifier
RF = RandomForestClassifier(random_state=seed)
hyperparameters = {
    'n_estimators': [50, 100, 200],
    'min_samples_split': [2, 4, 6],
    'min_samples_leaf': [2, 4, 6],
    'max_depth': [4, 6, 8, 10]
}
k_fold = KFold(n_splits=10, shuffle=True, random_state=seed)
grid_search = GridSearchCV(RF,
                        scoring='accuracy',
                        param_grid=hyperparameters,
                        cv=k_fold,
                        verbose=1
)
grid_search.fit(X_train,Y_train)
best_tunned = {}
best_tunned["model"] = grid_search.best_estimator_
best_tunned["score"] = grid_search.best_score_*100
print(best_tunned)

Fitting 10 folds for each of 108 candidates, totalling 1080 fits
{'model': RandomForestClassifier(max_depth=10, min_samples_leaf=6, n_estimators=50,
                       random_state=7), 'score': np.float64(80.28291909042832)}


### 2.2 Đáng giá mô hình với tập test

In [9]:
acc = best_tunned["model"].score(X_validation,Y_validation)*100
print("Score of Model", best_tunned["model"], ":", acc)

Score of Model RandomForestClassifier(max_depth=10, min_samples_leaf=6, n_estimators=50,
                       random_state=7) : 81.16883116883116


- Vậy RandomForest sau khi tunned có độ chuẩn xác (accuracy) ~81.17%